src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,163 @@
+                        var (syntax, symbol) = await FindDisposableLocalDeclaration(context.Document, context.Span, context.CancellationToken).ConfigureAwait(false);
+
+                        return syntax == null
+                            ? context.Document
+                            : await IntroduceUsingStatementAsync(context.Document, syntax, symbol, cancellationToken).ConfigureAwait(false);
+                    }));
+            }
+        }
+
+        private static async Task<(TLocalDeclarationSyntax, ILocalSymbol)> FindDisposableLocalDeclaration(Document document, TextSpan selection, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var declarationSyntax = root.FindNode(selection)?.FirstAncestorOrSelf<TLocalDeclarationSyntax>();
+            if (declarationSyntax == null)
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (!(semanticModel.GetOperation(declarationSyntax, cancellationToken) is IVariableDeclarationGroupOperation operation)
+                 || operation.Declarations.Length != 1)
+            {
+                return default;
+            }
+
+            var localDeclaration = operation.Declarations[0];
+            if (localDeclaration.Declarators.Length != 1)
+            {
+                return default;
+            }
+
+            var declarator = localDeclaration.Declarators[0];
+            if (declarator.Initializer == null)
+            {
+                return default;
+            }
+
+            var localType = declarator.Symbol?.Type;
+            if (localType == null)
+            {
+                return default;
+            }
+
+            if (declarator.Initializer.Value?.Syntax != null
+                && !await CodeRefactoringHelpers.RefactoringSelectionIsValidAsync(
+                    document,
+                    selection,
+                    node: declarationSyntax,
+                    holes: ImmutableArray.Create(declarator.Initializer.Value.Syntax),
+                    cancellationToken).ConfigureAwait(false))
+            {
+                return default;
+            }
+
+            var disposableType = semanticModel.Compilation?.GetSpecialType(SpecialType.System_IDisposable);
+            if (disposableType == null || !localType.InheritsFromOrEquals(disposableType, includeInterfaces: true))
+            {
+                return default;
+            }
+
+            return (declarationSyntax, declarator.Symbol);
+        }
+
+        protected static SyntaxNode FindSiblingContainingLastUsage(
+            SyntaxNode declarationSyntax,
+            ILocalSymbol localVariable,
+            SemanticModel semanticModel,
+            ISyntaxFactsService syntaxFactsService,
+            CancellationToken cancellationToken)
```

Comment:
1. You might want this be strongly typed on a TStatementSyntax type.  
2 Probably should be named "FindSiblingStatement..."